### ISSUE TITLE
Write RocksDB High Level Documentation

### ISSUE THREAD

**Issue by mvandeberg** [steemit/steem#2158]
We need some high level documentation on RocksDB integration with steemd.

**Comment by theoreticalbts** [comment]
Sorry for the wall of text, but there are a lot of issues that need to be addressed, if we're really serious about implementing this.

## Delayed node notes

You write:  *volatile state shall appear only on witness nodes, which by design can produce bad block (and fork) what next can result in returning back to the main blockchain branch and reverting the state to one of previous blocks. Other nodes (non-witnesses) by design shall accept only irreversible blocks*

What you are describing we call a *delayed node*.  The delayed node only pushes irreversible blocks.  The problem is, it needs some information about which blocks are irreversible.  There are only two
ways to know whether block B is irreversible:

- (a) You evaluate blocks, and check that a majority of witnesses have built valid blocks on block B.
- (b) Someone you trust tells you block B irreversible.

Option (a) requires you to run a non-delayed node anyway, since (by definition) the only way to check whether blocks are valid is to fully evaluate them.  Most users don't trust anybody, so they can't use option (b) [1].

Delayed nodes may theoretically be useful for very limited use cases, but most users will run non-delayed nodes.  The correct functioning of the p2p network depends on running non-delayed nodes.

[1] Just in case someone wants to use option (b), at one point in time we had a plugin called the *delayed node*, which rather than connecting to the p2p network, instead connected to another node and requested blocks over RPC, but only up to the last irreversible block.

I'm not sure this plugin was used by anyone, since it was removed in a refactor and AFAIK no one has asked for it back.  A delayed node can be implemented using a Python script to read LIB and transmit blocks up to LIB to another node.

## Fork handling performance issues

You write:  *Reverting the state changes up to specified block number* involves "re-evaluating all blocks starting from LastIrreversibleBlock + 1 up to ReferenceBlock." What is the performance in the usual case?  Here are the parameters for what I consider typical case:

- LIB is ~15 blocks behind head block.
- Each block updates 200 objects.

How many object updates does this cause?

- Replaying from LIB therefore requires replaying 14 blocks (2800 object updates).
- Two additional blocks on the new longest fork raises the total to 3200 object updates.

How does the current Steem / chainbase do it?

- Pop a single session for the last block, causing 200 object updates.
- Two additional blocks on the new longest fork raises the total to 600 object updates.

Deep forks are rare, shallow forks are common.

## No transaction rollback support

A transaction may fail after writing some state.  Many of our evaluators are written with the expectation that an `fc::exception` at any point will cause any prior writes to be undone.  In BitShares, we at first tried to write operations in two steps:  A first step `evaluate()` which checks for failure conditions but does not modify state, and a second step `apply()` which modifies all state.  This proved to be very impractical:

- A lot of code is common to the `evaluate()` and `apply()` functions
- Can't use the DB to detect violation of uniqueness constraints, since it only fails during `apply()`
- Multiple `operation` in the same `transaction` means you still need to undo if e.g. the last one fails

So in Steem we decided to just have `apply()`.

There needs to be a way to undo a partially completed transaction.

Going all the way back to LIB and replaying in case of a failed transaction, as well as a failed block, is far too much work.  (See above section.)

## No pending transaction rollback support

The node's current state is always `genesis + blocks + pending_transactions`.  This means state transitions need to be implemented like this:

```
push_tx( new_tx )
{
   // Pseudocode of algorithm implemented by _push_transaction

   auto session = start_undo_session();
   bool success = apply_transaction( new_tx );
   if( success )
      session.squash();
   else
      session.undo();
}

handle_tx( new_tx )
{
   // State is genesis + blocks + pending_transactions
   db.push_transaction( new_tx );
   // State is now genesis + blocks + pending_transactions + new_tx
}

handle_block( new_block )
{
   // On entry, state is genesis + blocks + pending_transactions

   vector< transaction > temp = db.pending_transactions;
   db.rollback_pending_transactions();
   // State is now genesis + blocks

   db.push_block( new_block );
   // State is now genesis + blocks + new_block

   for( auto& tx : temp )
      db.push_transaction( tx );
   // State is now genesis + blocks + new_block + pending_transactions
}
```

In Steem this logic is handled by the `without_pending_transactions()` function.

In more detail, Steem / Chainbase implements a list of undo states (you can think of it as a stack or queue).

- One undo state for each block.
- One undo state for successful pending transactions.
- When a new transaction is being pushed, one undo state for the current transaction.

When a transaction is pushed, a new undo state is created for it.  If the transaction fails, the undo state is popped (rollback).  If the transaction succeeds, the undo state is merged with the pending transactions state (so the new pending transaction state is effectively `pending_transactions + new_transaction`).

## No support for querying indexes

To determine the effects and validity of transactions and blocks, queries on any index need to be done.  So we need query methods which merge the query results from irreversible state and volatile state.

Let's say Alice has 100 Steem.  Alice sends 60 Steem to Bob, then Alice sends 70 Steem to Charlie in a short time (say, the following block).  A correctly functioning blockchain should permit the first transfer, then forbid the second.

What happens if we query irreversible state?

- Alice has 100 Steem in the irreversible state.
- Alice's account is not present in the volatile state.
- Alice sends 60 Steem to Bob.
- The query checks the irreversible state, and it shows that Alice has 100 Steem.
- Since Alice has enough Steem, the first transfer occurs.
- Alice now has 40 Steem in the volatile state.
- Alice still has 100 Steem in the irreversible state, because her transfer to Bob is later than LIB.
- Alice sends 70 Steem to Charlie.
- The query checks the irreversible state, and it shows that Alice has 100 Steem.
- Since Alice has enough Steem, the second transfer occurs.
- This is an error!

What happens if we query volatile state?

- Alice has 100 Steem in the irreversible state.
- Alice's account is not present in the volatile state.
- Alice sends 60 Steem to Bob.
- The query checks the volatile state, and it shows that Alice does not have an account.
- Since Alice does not exist, the first transfer fails.
- This is an error!

We need to be able to query a "merge" of the volatile state and irreversible state.  This merged state needs to support the same semantics as keeping in a single state, including:

- Existence.  Alice might exist in the irreversible state but not the volatile state (if she was not recently modified).  Alice might exist in the volatile state but not the irreversible state (if she was recently created).  Alice might exist in both states (if she was recently modified).  Or she might exist in neither state (if she really doesn't exist).
- Uniqueness.  Suppose we rely on uniqueness constraint to enforce business logic.  If `alice` account exists in irreversible state, but not in volatile state, then creating another account named `alice` must fail *even if* the name `alice` is unique in the volatile state.
- Iteration.  Some queries iterate through things (for example market orders by price, to match the best order).  The iteration logic must look for the next object in both containers and return the smaller one.  It must also handle cases of deletion and modification.
- Deletion.  Case (a): "Market order existed in the irreversible state, but was deleted in the volatile state."   Case (b): "Market order existed in the irreversible state, and hasn't been changed in the volatile state."  In Case (a), an operation that references the order (such as a cancel) should fail.  In Case (b), an operation that references the order should succeed.  The implementation needs to represent Case (a) and Case (b) differently, because it is expected to have different behavior in these cases.

## Partial merge of volatile state into irreversible state

The volatile state is *never* merged wholesale into the irreversible state.  How exactly does the merge occur, given that only part of the volatile state becomes irreversible at a time?  There are no details on this.

## Operations needed

Basically, the Steem design deeply bakes in expectation that you to have the left-to-right data structure of sessions, which supports:

- Commit sessions from the left
- Create sessions on the right
- Undo sessions from the right
- Squash the leftmost two sessions
- Query / write the cumulative state of all sessions with read-after-write consistency (including iteration and uniqueness on all indexes)

You can emulate all these functions.  Or you can refactor Steem to not require them.  The paper doesn't really go into enough depth.  I get the impression that there are a lot of complexities that haven't been considered.

It's an ambitious project.  The paper is a starting point, but I still have more questions than answers.

**Comment by vogel76** [comment]
1. Delayed nodes notes.
 * AFAIK the block is irreversible after producing its 11 successors, what means that major of wittnesses confirmed that block. So to determine block kind, its number shall be compared against blockchain head block and all blocks having numbers less than HeadBlockNo - 11 are irreversible.
 * delayed nodes are natural candidates to nodes serving API request, since they provide data produced on confirmed blockchain data. Actually I think current implementation can return bad information, since the state exactly holds all changes applied during processed transactions, which at most will be rolled back during processing undo.
 * could you elaborate why p2p network needs to operate on non-delayed node ?
2. Fork performance issues.
 * you forgot to mention that current implementation introduces serious overhead during each made modification by forcing whole object copies done just before the modification. Each time when any part of the object is modified. Also objects to be removed have to be copied into undo state data structures. This paradigm also seriously (and wrong) influenced object model design which is in very cases damaged because of object-copy-problem (i.e. objects can't store containers since copy will be slow, bigger objects are split into smallers etc).
New implementation does not need such copy at all.
 * During rebuilding the Volatile State we need to reevaluate very limited number of blocks (at most 11), directly operating in memory without any copies. At the moment dramatically damaged (mostly by using boost::interprocess::allocator and huge memory needs) steemd (configured as full-node and all enabled plugins) is able to process ~80 blocks per second - see replay speed chart:
https://docs.google.com/spreadsheets/d/1B3rxdZuvTJn2SCi26eto2IfD8rFMEjXlxHW04MwI1rU/edit?usp=sharing
what means that in current (dramatically slow) version it could take at most 1/8s.
The witness node currently can process ~650 blocks per second. So replaying the Volatile State will take 11/650 s (in current implementation).
Because new implementation will operate directly in-RAM and will have significantly reduced mem usage, block processing speed shall be much better.
Also number of data held in VolatileState is very limited (only changed objects) so also leads to reducing data access complexity in compare to current version which needs all data to reside in RAM, because otherwise it does not work at all.

**Comment by mvandeberg** [comment]
> could you elaborate why p2p network needs to operate on non-delayed node ?

If we are throwing out the existing p2p code in this refactor, then I suppose we can do anything.

With the current p2p code, peers only relay blocks/transactions that they apply and approve. Actually, having the node's default operation set to irreversible only is expensive. Consider the worst case scenario, a 21-way fork. Every witnesses is only producing on their own blocks. The p2p code needs to buffer all of the blocks because they could all be on the majority fork. And when one is made irreversible, how do you know which one? In fact, you have to push all of the blocks to attempt to make the longest chain, pushing, undoing, etc. until you find the longest chain. To simply push 1 block, you need to do the work of pushing all of them. In the best case scenario you have the same complexity. Blocks come in one after another with no forks. If you assume they are all valid, you know when a block becomes irreversible. In the worst case, the complexity for pushing one block is now O(n^2). In actuality, this is 21^2. Based on your cited value of ~650 blocks per second, a node you handle this, but we would be willing using 2/3 of the available performance for this algorithm.

That is just what needs to be done for a node to push blocks. The other problem is that many transactions have TaPoS set to reversible blocks (having it set to irreversible blocks somewhat defeats the point). The transaction cannot be validated by a node that only has reversible state. Your options are either a) apply pending state which as seen above can be prohibitively expensive or b) broadcast without validation which can open up the p2p network to spam and/or ddos.

I would also like to add that the merging of volatile state is needed for the modular blockchain fabric. It needs to be added one way or another. This has been the plan for over 6 months. I even outlined that on a whiteboard with Theoreticalbts many months back. Arguing that it is complicated so we shouldn't do it is invalid. In fact, the opposite is true. If RocksDB provides a better way to handle this merge on the fly than what we can implement on top of boost multi index, it would be a strong reason to use RocksDB.

**Comment by vogel76** [comment]
@youkaicountry I suppose this issue could be closed as the document has been created.
